photon-ml/src/main/scala/com/linkedin/photon/ml/avro/model/ModelProcessingUtils.scala

```diff
@@ -431,4 +459,74 @@ object ModelProcessingUtils {
     val colLatentFactors = loadLatentFactorsFromHDFS(colLatentFactorsPath.toString, sparkContext)
     new MatrixFactorizationModel(rowEffectType, colEffectType, rowLatentFactors, colLatentFactors)
   }
+
+  /**
+   * Save model metadata to a JSON file.
+   *
+   * @param outputDir The HDFS directory that will contain the metadata file
+   * @param params The model parameters, from which metadata is extracted
+   * @param sparkContext The Spark context
+   */
+  def saveGameModelMetadataToHDFS(outputDir: String, params: Params, sparkContext: SparkContext): Unit = {
+
+    // 2 ancillaries that decompose the problem into more manageable tasks
+    def writeConfigToJson(configs: Map[String, GLMOptimizationConfiguration]) =
+      configs
+        .map { case (cfgName, cfg) => s"""{ "name": "$cfgName", "configuration": ${cfg.toJson} }""" }
+        .mkString(",")
+
+    def writeConfigsArrayToJson(which: String, configsArray: Array[Predef.Map[String, GLMOptimizationConfiguration]]) =
+      s"""
+         |{ "configurations": "$which",
+         |  "values": [
+         |    ${configsArray.map(cfgs => writeConfigToJson(cfgs)).mkString(",")}
+         |  ]
+         |}
+       """.stripMargin
+
+    val feConfigurations = writeConfigsArrayToJson("fixed-effect", params.fixedEffectOptimizationConfigurations)
+    val reConfigurations = writeConfigsArrayToJson("random-effect", params.randomEffectOptimizationConfigurations)
+
+    IOUtils.toHDFSFile(sparkContext, outputDir + "model-metadata.json") {
+      writer => writer.println(
+        s"""
+           |{ "modelType": "${params.taskType}",
+           |  "modelName": "${params.applicationName}",
+           |  "modelVersion": "1.0",
+           |  "fixedEffectOptimizationConfigurations": $feConfigurations,
+           |  "randomEffectOptimizationConfigurations": $reConfigurations
+           |}
+         """.stripMargin)
+    }
+  }
+
+  /**
+   * Load model metadata from JSON file.
+   *
+   * NOTE: for now, we just output model type.
+   * TODO: load (and save) more metadata, and return an updated Params
+   *
+   * NOTE: if using the builtin Scala JSON parser, watch out, it's not thread safe!
+   *
+   * @param inputDir The HDFS directory where the metadata file is located
+   * @param sparkContext The Spark context
+   * @return A new copy of params with updates from the metadata file
+   */
+  def loadGameModelMetadataFromHDFS(inputDir: String, sparkContext: SparkContext): Params = {
+
+    val inputPath = new Path(inputDir + "model-metadata.json")
+    val fs = inputPath.getFileSystem(sparkContext.hadoopConfiguration)
+    val stream = fs.open(inputPath)
+    val fileContents = Source.fromInputStream(stream).getLines.mkString
+
+    val params = new Params
+    val modelTypeRE = """"modelType"\s*:\s*"(.+?)"""".r
```

Comment:
Regular Expression. Since in our context it collides with Random Effect, I'll expand to RegularExpression - but that makes for a really long identifier...